### PR TITLE
Report File Module

### PR DESCRIPTION
**Pull request by csemiao**
Report file module implemented
- Minimized coupling to mvc module
- Generates a report file report_yyyyMMdd.txt containing a list of reports that can be filtered according to date range and cost center code.
- Report file is an ascii table of applicable requests (see docs for details)
- One new stored procedure
- One new endpoint: GET reports, and it's associated controller and service (reuses the requests repo layer)
- Tested and updated docs

### CODE REVIEW

File: src/main/java/com/coastcapitalsavings/mvc/controllers/ReportsController.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.time.DateTimeException;
+import java.util.Date;
+import java.util.List;
+
+/**
+ * Created by Chris on 2017-03-21.
```
Comment (nanstc) [inline src/main/java/com/coastcapitalsavings/mvc/controllers/ReportsController.java]: Let's take your personal signature out ;)

Comment (nanstc) [inline src/main/java/com/coastcapitalsavings/mvc/controllers/ReportsController.java]: Let's take your personal signature out ;)

File: src/main/java/com/coastcapitalsavings/mvc/controllers/ReportsController.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+ */
+@RestController
+@RequestMapping(value="/reports")
+public class ReportsController {
+
+    @Autowired ReportService reportService;
+    @Autowired RequestService requestService;
+
+    // TODO: this endpoint needs to be locked to only administrators
+    // TODO: add cost center filtering to this and to requests/ get
```
Comment (nanstc) [inline src/main/java/com/coastcapitalsavings/mvc/controllers/ReportsController.java]: Thought cost centre filtering is already implemented?

File: src/main/resources/schema.sql [hunk 3]
```diff
@@ -676,4 +676,36 @@ CREATE PROCEDURE req_category_insert
...
+# Created On:  	2017-3-23
+#---------------------------------------------------------------------------------------------
+^;
+DROP PROCEDURE IF EXISTS req_request_lookUpByDateRangeAndCostCenterIncludeProducts^;
+CREATE PROCEDURE req_request_lookUpByDateRangeAndCostCenterIncludeProducts
+  (IN in_fromDate DATETIME,
+   IN in_toDate DATETIME,
+   IN in_costCenterCode VARCHAR(10))
+  BEGIN
+    SELECT r.requestId, r.dateCreated, r.submittedby, r.lastModified, r.lastModifiedBy, r.statusCode AS reqStat, pr.productCode, pr.statusCode AS prodStat
```
Comment (nanstc) [inline src/main/resources/schema.sql]: SQL too long, let's break it down a bit

File: src/main/resources/schema.dev [hunk 4]
```diff
@@ -698,4 +698,37 @@ CREATE PROCEDURE req_category_insert
...
+# Created On:  	2017-3-23
+#---------------------------------------------------------------------------------------------
+DELIMITER $$
+DROP PROCEDURE IF EXISTS req_request_lookUpByDateRangeAndCostCenterIncludeProducts $$
+CREATE PROCEDURE req_request_lookUpByDateRangeAndCostCenterIncludeProducts
+	(IN in_fromDate DATETIME,
+	 IN in_toDate DATETIME,
+	 IN in_costCenterCode VARCHAR(10))
+	BEGIN
+		SELECT r.requestId, r.dateCreated, r.submittedby, r.lastModified, r.lastModifiedBy, r.statusCode AS reqStat, pr.productCode, pr.statusCode AS prodStat
```
Comment (nanstc) [inline src/main/resources/schema.dev]: SQL too long. Let's break it down a bit

File: src/main/java/com/coastcapitalsavings/mvc/controllers/ReportsController.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        SimpleDateFormat df = new SimpleDateFormat("yyyyMMdd");
+
+        String filename = ("report_" + df.format(new Date()) + ".txt");
+        String fileContents = reportService.generateReport(requests);
+        response.addHeader("Content-disposition", "attachment;filename="  + filename);
+        response.setContentType("text/plain");
+
+        ServletOutputStream out = response.getOutputStream();
+        out.write(fileContents.getBytes(Charset.forName("UTF-8")));
+        response.flushBuffer();
```
Comment (nanstc) [inline src/main/java/com/coastcapitalsavings/mvc/controllers/ReportsController.java]: gonna be stupidly picky here, but do you mind adding an isCommited assert after flushBuffer? Just so we're sure that stuff is flushed. Easier for debugging too.

Comment (csemiao) [inline src/main/java/com/coastcapitalsavings/mvc/controllers/ReportsController.java]: changes made.

**Pull request merged**